Update covariant returns - add support for pointers

janvorli opened the pull request:
This change adds support for managed, unmanaged and function pointers as
return values. It also adds related tests.

janvorli committed:
Update covariant returns - add support for pointers

This change adds support for managed, unmanaged and function pointers as
return values. It also adds related tests.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -29,6 +102,24 @@ public static bool IsCompatibleWith(this TypeDesc thisType, TypeDesc otherType)
+            if (thisType.IsFunctionPointer && otherType.IsFunctionPointer)
+            {
+                return IsMethodSignatureCompatibleWith(thisType, otherType);
+            }
```
We should add something along the lines of:

```csharp
if (otherType.IsByRef || otherType.IsPointer || otherType.IsFunctionPointer)
{
    return false;
}
```

Passing those to `CanCastTo` will cause an assert in `CanCastToInternal`. `CanCastTo` expects a boxable input. If we boxed a pointer, something is wrong.

davidwrighton replied:
+1

davidwrighton replied:
I think it should check not just otherType but also thisType. Once we've finished the checks written above, any of those sorts of types are a false result.

tannergooding replied:
For function pointers, will it take calling convention compatibility into account?

tannergooding replied:
Looks like https://github.com/dotnet/runtime/pull/37454/files#diff-ac8d470d9a85bb53d9a33af63b36fe45R87 and I just missed it my first pass

janvorli committed:
Reflect PR feedback

janvorli commented:
@MichalStrehovsky, @davidwrighton I've added the checks you have suggested both to the runtime and the crossgen2.  I have just realized I should add more tests to cover these cases, so I am going to do it here.

janvorli committed:
Added more tests for the negative cases

janvorli commented:
Added tests for the newly added checks.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -20,6 +20,79 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
+        /// Note - this is a simplified version of what's described in the ECMA spec and it considers
+        /// pointers to be method-signature-compatible-with only if the signatures are the same.
+        /// </summary>
+        static bool IsMethodSignatureCompatibleWith(TypeDesc fn1Ttype, TypeDesc fn2Type)
```
I would just inline this check into the place that calls this. It's rather odd to see an API that talks about method signature compatibility taking two `TypeDesc` parameters.

janvorli replied:
I would like to keep it in a separate method so that its name and comment can express the relation to the ECMA spec. And also match the runtime code where we actually have a separate function pointer type.

MichalStrehovsky replied:
I would at least add an assert that both types `.IsFunctionPointer`... but take this as a nit since this is a private method and I don't care that much.

Basically, I find it odd that I can do `IsMethodSignatureCompatibleWith(context.GetWellKnownType(WellKnownType.String), context.GetWellKnownType(WellKnownType.String))` and this method will return `true`.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -20,6 +20,79 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
+                return true;
+            }
+
+            TypeFlags e1 = GetVerificationTypeElementType(type1);
```
Nit: inlining these temporaries into the `return` statement will speed things up here (no need to compute the verification element type for both if `type1.IsPrimitive` is false.

janvorli replied:
Makes sense.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -20,6 +20,79 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
+        /// Get TypeFlags of the reduced type of a type.
+        /// The reduced type concept is described in ECMA 335 chapter I.8.7
+        /// </summary>
+        static TypeFlags GetReducedTypeElementType(TypeDesc type)
```
Nit: we follow the CoreFX coding guidelines that is always explicit about accessibility of methods/types.

```suggestion
        private static TypeFlags GetReducedTypeElementType(TypeDesc type)
```

janvorli replied:
Right, it was just an overlook, thanks for spotting it.

janvorli committed:
Reflect PR feedback

* Add explicit "private" to a couple of added methods in crossgen2
* Don't call GetVerificationTypeElementType for non-primitive types

Also fix a compilation error on Linux with switch statement not covering
all the cases. It was correct, as the remaining cases were properly
handled further, but the compiler warnings as error didn't like it.

janvorli committed:
Reflect PR feedback

Add assert checking that both parameters are function pointers into
IsMethodSignatureCompatibleWith
